File: scraping/dining_menu_scrape.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import NoSuchElementException, ElementNotInteractableException
from selenium.common.exceptions import TimeoutException
import time
import re
import sqlite3

conn = sqlite3.connect('dining_hall_data.db')
cursor = conn.cursor()

# Function to insert data into SQLite
def insert_data(instance_data):
    sql = """
    INSERT INTO scraped_data (dining_hall, meal_period, name, dietary_restrictions, calories, fats, carbs, protein, sugar, serving_size)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    """
    values = (
        instance_data['dining_hall'],
        instance_data['meal_period'],
        instance_data['name'],
        ','.join(instance_data['dietary_restrictions']),  # Convert list to comma-separated string
        instance_data['calories'],
        instance_data['fats'],
        instance_data['carbs'],
        instance_data['protein'],
        instance_data['sugar'],
        instance_data['serving_size']
    )
    # Execute the query with the given values
    cursor.execute(sql, values)
    conn.commit()  # Commit the transaction

# Function to perform the scraping
def scrape_data():
    # Set up WebDriver
    driver = webdriver.Chrome()

    # Open the UC Berkeley dining menus page
    url = 'https://dining.berkeley.edu/menus/'
    driver.get(url)

    # Create a list to store all data
    menu_data = []

    # Define a list of dining halls to scrape
    dining_halls = driver.find_elements(By.CLASS_NAME, 'cafe-title')

    # Get the meal periods (e.g., Breakfast, Lunch, Dinner)
    all_meal_periods = driver.find_elements(By.CLASS_NAME, 'preiod-name')

    # Loop over each dining hall
    for hall_index in range(len(dining_halls)):
        hall = dining_halls[hall_index]
        dining_hall_name = dining_halls[hall_index].text  # Extract the name of the dining hall

        # Close the modal if it's open before clicking the next dining hall
        try:
            close_button = WebDriverWait(driver, 10).until(
                EC.element_to_be_clickable((By.CLASS_NAME, 'cald-close'))
            )
            driver.execute_script("arguments[0].click();", close_button)  # Click via JavaScript
        except TimeoutException:
            pass

        driver.execute_script("arguments[0].scrollIntoView(true);", hall)  # Scroll to the element
        driver.execute_script("arguments[0].click();", hall)

        time.sleep(2)  # Wait for the page to load
        meals_per_day = len(all_meal_periods) // len(dining_halls)

        # Calculate the start and end indices for the current hall's meal periods
        start_idx = hall_index * meals_per_day
        end_idx = start_idx + meals_per_day

        # Get the meal periods specific to this dining hall
        meal_periods = all_meal_periods[start_idx:end_idx]

        # Loop over these meal periods
        for meal_period in meal_periods:
            meal_period_name = meal_period.text.split(' ')[-1]  # Extract the meal period
            try:
                time.sleep(1)
                WebDriverWait(driver, 15).until(EC.element_to_be_clickable(meal_period))
                driver.execute_script("arguments[0].scrollIntoView(true);", meal_period)
                driver.execute_script("arguments[0].click();", meal_period)

                time.sleep(2)  # Wait for the page to load

                # Extract all meal items for the current meal period
                meal_items = meal_period.find_elements(By.CLASS_NAME, 'recip')

                for meal_item in meal_items:
                    meal_name = meal_item.find_element(By.TAG_NAME, 'span').text

                    # Extract dietary icons
                    icons = meal_item.find_elements(By.CLASS_NAME, 'food-icon')
                    dietary_restrictions = [icon.find_element(By.TAG_NAME, 'img').get_attribute('alt') for icon in icons]

                    # Attempt to click the meal item using JavaScript
                    try:
                        driver.execute_script("arguments[0].scrollIntoView(true);", meal_item)
                        WebDriverWait(driver, 15).until(EC.visibility_of(meal_item))
                        driver.execute_script("arguments[0].click();", meal_item)

                        # Wait for the recipe card to load
                        serving_size_text = WebDriverWait(driver, 15).until(
                            EC.visibility_of_element_located((By.CLASS_NAME, 'title'))
                        ).find_element(By.CLASS_NAME, 'serving-size').text

                        serving_size = re.search(r'\d+', serving_size_text).group()

                        # Get the card containing all the nutrients
                        nutrient_card = WebDriverWait(driver, 15).until(
                            EC.visibility_of_element_located((By.CLASS_NAME, 'nutration-details'))
                        )
                        nutrient_list = nutrient_card.find_elements(By.TAG_NAME, 'li')

                        # Extract nutrient information by index
                        nutrients = {
                            'calories': nutrient_list[0].text.split(":")[-1].strip(),
                            'fats': nutrient_list[1].text.split(":")[-1].strip(),
                            'carbs': nutrient_list[6].text.split(":")[-1].strip(),
                            'protein': nutrient_list[9].text.split(":")[-1].strip(),
                            'sugar': nutrient_list[8].text.split(":")[-1].strip()
                        }

                        # Grab all nutrient data about specific instance
                        instance_data = {
                            'dining_hall': dining_hall_name,
                            'meal_period': meal_period_name,
                            'name': meal_name,
                            'dietary_restrictions': dietary_restrictions,
                            'calories': float(nutrients['calories']),  # Convert to integer
                            'fats': float(nutrients['fats']),          # Convert to integer
                            'carbs': float(nutrients['carbs']),        # Convert to integer
                            'protein': float(nutrients['protein']),    # Convert to integer
                            'sugar': float(nutrients['sugar']),        # Convert to integer
                            'serving_size': float(serving_size)        # Convert to integer
                        }

                        # Add nutrient info to the menu data
                        menu_data.append(instance_data)
                        # Insert data into the SQLite database
                        insert_data(instance_data)

                    except (NoSuchElementException, ElementNotInteractableException) as e:
                        print(f"Error loading nutrient details for {meal_name}: {e}")

                    time.sleep(1)  # Wait a bit before moving to the next meal item

            except ElementNotInteractableException as e:
                print(f"Element not interactable: {e}")

    # Close the browser
    driver.quit()
    print(menu_data)

# Call the scraping function
scrape_data()

File: scraping/dining_menu_scrape.js
const { Builder, By, until } = require('selenium-webdriver');
const sqlite3 = require('sqlite3').verbose();
const schedule = require('node-schedule');
const fs = require('fs').promises;

// Database connection
const db = new sqlite3.Database('dining_hall_data.db');

// Create a list to store all data
let menuData = [];

// Function to clear the table
function clearTable() {
    return new Promise((resolve, reject) => {
        db.run("DELETE FROM scraped_data", (err) => {
            if (err) {
                console.error(`Error clearing table: ${err}`);
                reject(err);
            } else {
                console.log("Table cleared successfully");
                resolve();
            }
        });
    });
}

// Function to insert data into SQLite
function insertData(instanceData) {
    return new Promise((resolve, reject) => {
        const sql = `
            INSERT INTO scraped_data (
                dining_hall, meal_period, name, dietary_restrictions, 
                calories, fats, carbs, protein, sugar, serving_size
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        `;
        const values = [
            instanceData.dining_hall,
            instanceData.meal_period,
            instanceData.name,
            instanceData.dietary_restrictions.join(','),
            instanceData.calories,
            instanceData.fats,
            instanceData.carbs,
            instanceData.protein,
            instanceData.sugar,
            instanceData.serving_size
        ];

        db.run(sql, values, (err) => {
            if (err) reject(err);
            else resolve();
        });
    });
}

// Function to perform the scraping
async function scrapeData() {
    let driver;
    try {
        await clearTable();
        driver = await new Builder()
            .forBrowser('chrome')
            .setChromeOptions(new (require('selenium-webdriver/chrome').Options)().addArguments('--headless=new'))
            .build();

        // Open the UC Berkeley dining menus page
        await driver.get('https://dining.berkeley.edu/menus/');

        // Define a list of dining halls to scrape
        const diningHalls = await driver.findElements(By.className('cafe-title'));
        const allMealPeriods = await driver.findElements(By.className('preiod-name'));

        // Loop over each dining hall
        for (let hallIndex = 0; hallIndex < diningHalls.length; hallIndex++) {
            const hall = diningHalls[hallIndex];
            const diningHallName = await diningHalls[hallIndex].getText();

            // Close modal if open
            try {
                const closeButton = await driver.wait(
                    until.elementLocated(By.className('cald-close')),
                    10000
                );
                await driver.executeScript("arguments[0].click();", closeButton);
            } catch (error) {
                // Ignore if modal isn't present
            }

            await driver.executeScript("arguments[0].scrollIntoView(true);", hall);
            await driver.executeScript("arguments[0].click();", hall);

            await new Promise(resolve => setTimeout(resolve, 2000));

            const mealsPerDay = Math.floor(allMealPeriods.length / diningHalls.length);
            const startIdx = hallIndex * mealsPerDay;
            const endIdx = startIdx + mealsPerDay;
            const mealPeriods = allMealPeriods.slice(startIdx, endIdx);

            for (const mealPeriod of mealPeriods) {
                const mealPeriodName = (await mealPeriod.getText()).split(' ').pop();
                
                try {
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    await driver.wait(until.elementIsEnabled(mealPeriod), 15000);
                    await driver.executeScript("arguments[0].scrollIntoView(true);", mealPeriod);
                    await driver.executeScript("arguments[0].click();", mealPeriod);

                    await new Promise(resolve => setTimeout(resolve, 2000));

                    const mealItems = await mealPeriod.findElements(By.className('recip'));

                    for (const mealItem of mealItems) {
                        const mealName = await mealItem.findElement(By.css('span')).getText();
                        const icons = await mealItem.findElements(By.className('food-icon'));
                        const dietaryRestrictions = await Promise.all(
                            icons.map(async icon => {
                                const img = await icon.findElement(By.css('img'));
                                return img.getAttribute('alt');
                            })
                        );

                        try {
                            await driver.executeScript("arguments[0].scrollIntoView(true);", mealItem);
                            await driver.wait(until.elementIsVisible(mealItem), 15000);
                            await driver.executeScript("arguments[0].click();", mealItem);

                            const servingSizeElement = await driver.wait(
                                until.elementLocated(By.css('.title .serving-size')),
                                15000
                            );
                            const servingSizeText = await servingSizeElement.getText();
                            const servingSize = servingSizeText.match(/\d+/)[0];

                            const nutrientCard = await driver.wait(
                                until.elementLocated(By.className('nutration-details')),
                                15000
                            );
                            const nutrientList = await nutrientCard.findElements(By.css('li'));

                            const nutrients = {
                                calories: await nutrientList[0].getText().then(text => text.split(':').pop().trim()),
                                fats: await nutrientList[1].getText().then(text => text.split(':').pop().trim()),
                                carbs: await nutrientList[6].getText().then(text => text.split(':').pop().trim()),
                                protein: await nutrientList[9].getText().then(text => text.split(':').pop().trim()),
                                sugar: await nutrientList[8].getText().then(text => text.split(':').pop().trim())
                            };

                            const instanceData = {
                                dining_hall: diningHallName,
                                meal_period: mealPeriodName,
                                name: mealName,
                                dietary_restrictions: dietaryRestrictions,
                                calories: parseFloat(nutrients.calories),
                                fats: parseFloat(nutrients.fats),
                                carbs: parseFloat(nutrients.carbs),
                                protein: parseFloat(nutrients.protein),
                                sugar: parseFloat(nutrients.sugar),
                                serving_size: parseFloat(servingSize)
                            };

                            menuData.push(instanceData);
                            await insertData(instanceData);

                        } catch (error) {
                            console.error(`Error loading nutrient details for ${mealName}:`, error);
                        }

                        await new Promise(resolve => setTimeout(resolve, 1000));
                    }
                } catch (error) {
                    console.error("Element not interactable:", error);
                }
            }
        }
    } catch (error) {
        console.error("Scraping error:", error);
    } finally {
        if (driver) {
            await driver.quit();
        }
        console.log(menuData);
    }
}

// Function to run the job
async function job() {
    console.log("Running the scraping task...");
    await scrapeData();
    
    // Write data to JSON file
    try {
        await fs.writeFile(
            "src/meal_data.json",
            JSON.stringify(menuData, null, 4)
        );
    } catch (error) {
        console.error("Error writing to JSON file:", error);
    }
}

// Run job immediately
job();

// Schedule job to run daily at 12:01 AM
schedule.scheduleJob('1 0 * * *', job); 